Implement terms and conditions agreement

richardpattinson opened the pull request:
### Context

Introduce a system prompt for users to agree to the service's terms and conditions upon their first sign-in and at periodic intervals thereafter to ensure ongoing compliance and user understanding of the service policies. The T&C document is ready for implementation.

Objectives:

Ensure Compliance: Guarantee that all users are aware of and agree to the terms and conditions governing the use of the service.

User Awareness: Enhance user understanding of service policies, including data use, privacy, and user responsibilities.

Periodic Acknowledgment: Require users to periodically review and agree to the terms and conditions to account for updates or changes.

### Link to Trello card

<!-- http://trello.com/123-example-card -->

### Checklist

- [x] Attach to Trello card
- [x] Rebased main
- [x] Cleaned commit history
- [x] Tested by running locally

richardpattinson committed:
Implement terms and conditions agreement

malcolmbaig commented on app/controllers/concerns/dsi_authenticatable.rb:
```diff
@@ -4,8 +4,11 @@ module DsiAuthenticatable
   included do
     before_action :authenticate_dsi_user!
     before_action :handle_expired_session!
+    before_action :enforce_terms_and_conditions_acceptance!
```
I don't think we should conflate authentication concerns with the terms and conditions flow. We could alternatively put T&Cs logic in the CheckRecordsController where it'll still apply to all CTR requests, and make the check only if `current_dsi_user` exists.

malcolmbaig commented on app/controllers/concerns/dsi_authenticatable.rb:
```diff
@@ -16,6 +19,12 @@ def authenticate_dsi_user!
+  def enforce_terms_and_conditions_acceptance!
+    if FeatureFlags::FeatureFlag.active?("terms_and_conditions") && current_dsi_user.terms_and_conditions_version_accepted != CURRENT_TERMS_AND_CONDITIONS_VERSION
+        redirect_to check_records_terms_and_conditions_path
+    end
```
This check needs updating so we re-prompt them every six months (or whatever value was agreed).

malcolmbaig commented on app/controllers/check_records/terms_and_conditions_controller.rb:
```diff
@@ -0,0 +1,14 @@
+    def update
+      current_dsi_user.update!(terms_and_conditions_version_accepted: CURRENT_TERMS_AND_CONDITIONS_VERSION,
+                              terms_and_conditions_timestamp: Time.zone.now)
+      redirect_to check_records_root_path notice: "Terms and conditions accepted"
```
```suggestion
      redirect_to check_records_root_path, notice: "Terms and conditions accepted"
```

malcolmbaig commented on config/routes/check_records.rb:
```diff
@@ -13,6 +13,9 @@
   post "/auth/developer/callback" => "omniauth_callbacks#dfe_bypass"
 
+  get "/terms-and-conditions", to: "terms_and_conditions#show"
+  patch "/auth/developer/callback" => "terms_and_conditions#update"
```
patch path looks incorrect here.

malcolmbaig commented on spec/system/check_records/user_accepts_terms_and_conditions_spec.rb:
```diff
@@ -0,0 +1,39 @@
+    then_i_am_signed_in
+    and_i_am_redirected_to_the_terms_and_conditions_page
+    then_i_click_accept
+    then_i_am_taken_to_the_root
```
Let's also assert that we see the flash notice (will validate the suggested change [here](https://github.com/DFE-Digital/access-your-teaching-qualifications/pull/679/files#r1647343820)), and confirm the database changes on the dsi_user record.

malcolmbaig reviewed (changes requested):
A few comments left inline. After they're addressed, would you mind testing this locally and adding screenshots to the PR description?

malcolmbaig commented on app/controllers/check_records/check_records_controller.rb:
```diff
@@ -16,5 +19,13 @@ class CheckRecordsController < ApplicationController
+      if current_dsi_user &&
+        (FeatureFlags::FeatureFlag.active?("terms_and_conditions") &&
+          current_dsi_user.terms_and_conditions_version_accepted != CURRENT_TERMS_AND_CONDITIONS_VERSION)
+          redirect_to check_records_terms_and_conditions_path
```
There's still no logical check here based on the datetime they accepted the agreement.

malcolmbaig commented on config/routes/check_records.rb:
```diff
@@ -18,6 +18,12 @@
+  patch "/terms-and-conditions" => "terms_and_conditions#update"
+
+  get "/search", to: "search#new"
+  get "/result", to: "search#show"
```
Looks like a merge conflict from the search results work.

malcolmbaig replied:
```suggestion
```

malcolmbaig commented on spec/system/check_records/user_accepts_terms_and_conditions_spec.rb:
```diff
@@ -0,0 +1,49 @@
+    then_i_click_accept
+    then_i_am_taken_to_the_root
+    and_i_see_the_successful_notification
+    and_the_dsi_user_has_been_updated
```
Once the timestamp check has also been added, this spec can be expanded to travel forward in time and assert that the user is prompted again.

malcolmbaig commented on app/controllers/check_records/check_records_controller.rb:
```diff
@@ -16,5 +19,24 @@ class CheckRecordsController < ApplicationController
+    end
+
+    def acceptance_expired
+      current_dsi_user.terms_and_conditions_timestamp > 12.months.ago
```
To confirm - it's 12 months and not 6?

malcolmbaig commented on spec/system/check_records/user_accepts_terms_and_conditions_spec.rb:
```diff
@@ -0,0 +1,65 @@
+    then_i_click_accept
+    then_i_am_taken_to_the_root
+    and_i_see_the_successful_notification
+    and_the_dsi_user_has_been_updated
```
nit: can remove the last step, not testing anything with its addition.

saliceti committed:
Create terraform environment for review apps on AKS

malcolmbaig committed:
Fix terms and conditions logic at session timeout

When a user session times out, the current terms and conditions
enforcement in CTR will result in an infinite redirect loop.

This is because if an expired session is detected, we send the user to
the SignOutController. Here the controller should clear the session and
send the user to the sign in page.

With the addition of the terms and conditions logic, the sign out
controller does the enforcement check in a before filter before it gets
the chance to clear the session.

To fix the redirect loop this creates:

- Add the sign in and sign out paths for CTR to the list of paths the
  terms and conditions enforcement skips

This commit also includes minor refactors and reformatting carried out
during investigation of the bug.

felixclack commented on db/schema.rb:
```diff
@@ -117,6 +117,8 @@
     t.datetime "created_at", null: false
     t.datetime "updated_at", null: false
+    t.string "terms_and_conditions_version_accepted"
+    t.datetime "terms_and_conditions_timestamp"
```
Rails conventions are to use an `_at` suffix for timestamp fields. How about applying that here?

eg.

```suggestion
    t.datetime "terms_and_conditions_accepted_at"
```

felixclack commented on Gemfile.lock:
```diff
@@ -276,7 +276,7 @@ GEM
       faraday (~> 2.0)
       faraday-follow_redirects
-    jwt (2.8.1)
+    jwt (2.8.2)
```
Looks like some unrelated changes have snuck in here.

felixclack commented on app/controllers/check_records/check_records_controller.rb:
```diff
@@ -16,5 +18,27 @@ class CheckRecordsController < ApplicationController
+
+    def acceptance_expired
+      current_dsi_user.terms_and_conditions_timestamp < 12.months.ago
+    end
```
This is business logic and would be a good fit for being implemented within a model.

Either as a method on user or as an object that accepts the relevant fields and returns a boolean.

eg.

```
# app/models/user.rb
class User
  ...
  
  CURRENT_TERMS_AND_CONDITIONS_VERSION = "1.0".freeze

  def acceptance_required?
    !current_version_accepted? || acceptance_expired?
  end
  
  private

  def current_version_accepted?
    terms_and_conditions_version_accepted == CURRENT_TERMS_AND_CONDITIONS_VERSION
  end

  def acceptance_expired?
    terms_and_conditions_at.before?(12.months.ago)
  end
end
```

Or as a separate model
```
# app/models/terms_expiry.rb
class TermsExpiry
  include ActiveModel::Model
  CURRENT_TERMS_AND_CONDITIONS_VERSION = "1.0".freeze

  attr_accessor :accepted_at, :version

  validates :accepted_at, presence: true
  validates :version, presence: true

  def acceptance_required?
    !current_version_accepted? || acceptance_expired
  end

  private

  def current_version_accepted?
    version == CURRENT_TERMS_AND_CONDITIONS_VERSION
  end

  def acceptance_expired?
    accepted_at.before?(12.months.ago)
  end
end
```

richardpattinson replied:
I've put the logic in the dsi_users model

felixclack commented on app/controllers/check_records/terms_and_conditions_controller.rb:
```diff
@@ -0,0 +1,16 @@
+      current_dsi_user.update!(
+        terms_and_conditions_version_accepted: CURRENT_TERMS_AND_CONDITIONS_VERSION,
+        terms_and_conditions_timestamp: Time.zone.now
+      )
```
I would define a method for accepting terms and keep this logic out of the controller.

eg. `current_dsi_user.accept_terms!`

felixclack commented on app/views/check_records/terms_and_conditions/show.html.erb:
```diff
@@ -0,0 +1,210 @@
+
+<div class="govuk-grid-row">
+  <div class="govuk-grid-column-two-thirds">
+    <h1 class="govuk-heading-l">Terms and Conditions    </h1>
```
```suggestion
    <h1 class="govuk-heading-l">Terms and Conditions</h1>
```

felixclack commented on app/views/check_records/terms_and_conditions/show.html.erb:
```diff
@@ -0,0 +1,210 @@
+    <ol type="a" class="govuk-body">
+        <li class="govuk-body">our privacy notice that details the processing of Personal Data through the Service (as may be amended by us from time to time);</li>
+
+        <li class="govuk-body">any <a class="govuk-link" href="https://interactions.signin.education.gov.uk/terms">DfE Sign-In</a> terms and conditions that govern the access to the Service (as such terms of use may be amended by us from time to time).</li>
```
```suggestion
      <li class="govuk-body">our privacy notice that details the processing of Personal Data through the Service (as may be amended by us from time to time);</li>

      <li class="govuk-body">any <a class="govuk-link" href="https://interactions.signin.education.gov.uk/terms">DfE Sign-In</a> terms and conditions that govern the access to the Service (as such terms of use may be amended by us from time to time).</li>
```

felixclack commented on app/views/check_records/terms_and_conditions/show.html.erb:
```diff
@@ -0,0 +1,210 @@
+    <ol type="a" class="govuk-body">
+        <li class="govuk-body">your use of data obtained from the Service complies with Data Protection Legislation; and</li>
+
+        <li class="govuk-body">any processors required to access the Service on your organisation’s behalf do so with a valid basis for processing and always use any Personal Data obtained from the Service in compliance with Data Protection Legislation and in accordance with these Terms.</li>
```
```suggestion
      <li class="govuk-body">your use of data obtained from the Service complies with Data Protection Legislation; and</li>

      <li class="govuk-body">any processors required to access the Service on your organisation’s behalf do so with a valid basis for processing and always use any Personal Data obtained from the Service in compliance with Data Protection Legislation and in accordance with these Terms.</li>
```

felixclack commented on app/views/check_records/terms_and_conditions/show.html.erb:
```diff
@@ -0,0 +1,210 @@
+    <p class="govuk-body">3.4 As a user of this Service, You will at all times:</p>
+    <ol type="a" class="govuk-body">
+
+        <li class="govuk-body">comply with the provisions of Data Protection Legislation in respect of all Personal Data, understanding that such Personal Data is provided to enable only authorised users from a registered organisation to carry out necessary checks on whether a trainee or teacher has:</li>
```
```suggestion
      <li class="govuk-body">comply with the provisions of Data Protection Legislation in respect of all Personal Data, understanding that such Personal Data is provided to enable only authorised users from a registered organisation to carry out necessary checks on whether a trainee or teacher has:</li>
```

felixclack commented on app/views/check_records/terms_and_conditions/show.html.erb:
```diff
@@ -0,0 +1,210 @@
+                <li class="govuk-body">successfully completed a leadership or specialist national professional qualification;</li>
+
+                <li class="govuk-body">any teaching restrictions placed against them or has been the subject of a decision by the Secretary of State for Education not to impose a prohibition order following a determination by a professional conduct panel of unacceptable professional conduct or conduct that may bring the teaching profession into disrepute or the conviction of a relevant offence.</ ...
+            </ol>
```
```suggestion
    <ol type="i" class="govuk-body">
      <li class="govuk-body">achieved qualified teacher status, early years teacher status or early years professional status;</li>

      <li class="govuk-body">completed their induction;</li>

      <li class="govuk-body">been awarded a mandatory qualification for teachers of hearing impaired or visually impaired pupils;</li>

      <li class="govuk-body">successfully completed a leadership or specialist national professional qualification;</li>

      <li class="govuk-body">any teaching restrictions placed against them or has been the subject of a decision by the Secretary of State for Education not to impose a prohibition order following a determination by a professional conduct panel of unacceptable professional conduct or conduct that may bring the teaching profession into disrepute or the conviction of a relevant offence.</li>
```

felixclack commented on app/views/check_records/terms_and_conditions/show.html.erb:
```diff
@@ -0,0 +1,210 @@
+
+        <li class="govuk-body">reserves the right to reset user passwords.</li>
+
+        <li class="govuk-body">owns all intellectual property rights (which is protected by Crown copyright) in and to the Service which includes any and all data (including Personal Data) or other information or material made available or published via the Service and these rights are licensed (not sold) to You. You have no intellectual property rights in relation to the foregoing other than the ...
```
```suggestion
      <li class="govuk-body">takes no responsibility for the accuracy or completeness or the security of any personal data transferred to, used in or stored on your system or in your local files. In such circumstances, for the purposes of Data Protection Legislation, You understand that your organisation is the controller for any such personal data and any activities carried out with this data must comply with Data Protection Legislation.</li>

      <li class="govuk-body">takes reasonable steps to ensure the quality, accuracy, and completeness of its data but accept that these factors cannot be guaranteed.</li>

      <li class="govuk-body">reserves the right to reset user passwords.</li>

      <li class="govuk-body">owns all intellectual property rights (which is protected by Crown copyright) in and to the Service which includes any and all data (including Personal Data) or other information or material made available or published via the Service and these rights are licensed (not sold) to You. You have no intellectual property rights in relation to the foregoing other than the right to use them in accordance with these Terms.</li>
```

felixclack commented on app/views/check_records/terms_and_conditions/show.html.erb:
```diff
@@ -0,0 +1,210 @@
+
+    <li class="govuk-body">seek prior written advice from DfE where You are unsure whether a proposed use of the Service is permitted under Data Protection Legislation.</li>
+
+    <li class="govuk-body">indemnify and hold harmless DfE (and for the avoidance of doubt, the TRA) against any and all losses, liabilities, demands, fines, penalties, claims for compensation, costs (including without limitation legal costs), expenses or damages arising from any unauthorised use, loss or corruption of Personal Data available via the Service where the foregoing arises out of or i ...
```
```suggestion
      <li class="govuk-body">not disclose any Personal Data from the Service to any unauthorised third party.</li>

      <li class="govuk-body">abide by all instructions, conditions and other requirements made known to You by us when accessing and using the Service.</li>

      <li class="govuk-body">seek prior written advice from DfE where You are unsure whether a proposed use of the Service is permitted under Data Protection Legislation.</li>

      <li class="govuk-body">indemnify and hold harmless DfE (and for the avoidance of doubt, the TRA) against any and all losses, liabilities, demands, fines, penalties, claims for compensation, costs (including without limitation legal costs), expenses or damages arising from any unauthorised use, loss or corruption of Personal Data available via the Service where the foregoing arises out of or in connection with any act or omission You have taken.</li>
```

felixclack commented on app/views/check_records/terms_and_conditions/show.html.erb:
```diff
@@ -0,0 +1,210 @@
+
+        <li class="govuk-body">any automated analytical technique aimed at analysing content in digital form to generate information which includes but is not limited to patterns, trends and correlations,</li>
+
+        <li class="govuk-body">the provisions in this Clause should be treated as an express reservation of our rights in this regard, including for the purposes of Article 4(3) of Digital Copyright Directive ((EU) 2019/790). This Clause shall not apply insofar as (but only to the extent that) we are unable to exclude or limit text or data mining or web scraping activity by contract under the law ...
```
```suggestion
      <li class="govuk-body">any "robot", "bot", "spider", "scraper" or other automated device, program, tool, algorithm, code, process or methodology to access, obtain, copy, monitor or republish any portion of the Service or any content or services accessed via the same; or</li>

      <li class="govuk-body">any automated analytical technique aimed at analysing content in digital form to generate information which includes but is not limited to patterns, trends and correlations,</li>

      <li class="govuk-body">the provisions in this Clause should be treated as an express reservation of our rights in this regard, including for the purposes of Article 4(3) of Digital Copyright Directive ((EU) 2019/790). This Clause shall not apply insofar as (but only to the extent that) we are unable to exclude or limit text or data mining or web scraping activity by contract under the laws which are applicable to us.</li>
```

felixclack commented on app/views/check_records/terms_and_conditions/show.html.erb:
```diff
@@ -0,0 +1,210 @@
+
+    <p class="govuk-body"><a class="govuk-link" mailto="teaching.status@education.gov.uk">Email:	teaching.status@education.gov.uk</a></p>
+
+    <p class="govuk-body">Website:	<a class="govuk-link" href="https://www.gov.uk/government/organisations/teaching-regulation-agency">https://www.gov.uk/government/organisations/teaching-regulation-agency</a></p>
```
```suggestion
    <p class="govuk-body">Website: <a class="govuk-link" href="https://www.gov.uk/government/organisations/teaching-regulation-agency">https://www.gov.uk/government/organisations/teaching-regulation-agency</a></p>
```

felixclack commented on app/views/check_records/terms_and_conditions/show.html.erb:
```diff
@@ -0,0 +1,210 @@
+
+    <p class="govuk-body">Please contact us if You have any queries regarding these Terms or your use of the Service via our contact details:</p>
+
+    <p class="govuk-body"><a class="govuk-link" mailto="teaching.status@education.gov.uk">Email:	teaching.status@education.gov.uk</a></p>
```
```suggestion
    <p class="govuk-body"><a class="govuk-link" mailto="teaching.status@education.gov.uk">Email: teaching.status@education.gov.uk</a></p>
```

felixclack commented on app/views/check_records/terms_and_conditions/show.html.erb:
```diff
@@ -0,0 +1,210 @@
+          <li class="govuk-body">successfully completed a leadership or specialist national professional qualification;</li>
+
+          <li class="govuk-body">any teaching restrictions placed against them or has been the subject of a decision by the Secretary of State for Education not to impose a prohibition order following a determination by a professional conduct panel of unacceptable professional conduct or conduct that may bring the teaching profession into disrepute or the conviction of a relevant offence.</li>
+        </ol>
```
Is this supposed to be a separate list? It is invalid HTML in its current form.

felixclack replied:
This is invalid HTML and still needs addressing.

felixclack commented on app/views/check_records/terms_and_conditions/show.html.erb:
```diff
@@ -0,0 +1,210 @@
+    </ol>
+    <p class="govuk-body">3.5 As a user of this Service, You understand DfE:</p>
+    <ol type="a">
+        <li class="govuk-body">takes no responsibility for the accuracy or completeness or the security of any personal data transferred to, used in or stored on your system or in your local files. In such circumstances, for the purposes of Data Protection Legislation, You understand that your organisation is the controller for any such personal data and any activities carried out with this data  ...
```
Indentation looks inconsistent here compared to the rest of the file. It is using 4 spaces whereas we use 2 spaces everywhere else.

felixclack commented on app/controllers/check_records/check_records_controller.rb:
```diff
@@ -16,5 +18,16 @@ class CheckRecordsController < ApplicationController
+    def enforce_terms_and_conditions_acceptance!
+      return if request.path == check_records_terms_and_conditions_path
+      return if request.path == check_records_sign_in_path
+      return if request.path == check_records_sign_out_path
```
Putting this routing logic here means the method is violating the single responsibility principle. I would consider moving that logic to a skip_before_action or conditionally include this where relevant.

Also, this code is untested, let's add a test for it.

richardpattinson replied:
Ive added some lines that check that we cannot navigate away from the terms and conditions page and have moved these guard conditions into skip_before_actions on the relevant controllers. Is a seperate system spec needed for checking that sign_in and out are still possible without approving the t&c's?

felixclack commented on Gemfile.lock:
```diff
@@ -278,7 +278,7 @@ GEM
       faraday (~> 2.0)
       faraday-follow_redirects
-    jwt (2.8.2)
+    jwt (2.8.1)
```
This change appears to be unrelated to terms and conditions.

felixclack commented on app/controllers/check_records/check_records_controller.rb:
```diff
@@ -16,5 +18,12 @@ class CheckRecordsController < ApplicationController
+
+    def enforce_terms_and_conditions_acceptance!
+      if current_dsi_user &&
+        (FeatureFlags::FeatureFlag.active?("terms_and_conditions") && current_dsi_user.acceptance_required?)
```
Do we need the feature flag?

felixclack replied:
```suggestion
      if current_dsi_user&.acceptance_required?
```

felixclack commented on app/controllers/concerns/dsi_authenticatable.rb:
```diff
@@ -6,6 +6,8 @@ module DsiAuthenticatable
     before_action :handle_expired_session!
   end
 
+  CURRENT_TERMS_AND_CONDITIONS_VERSION = "1.0".freeze
```
This isn't needed anymore

felixclack commented on db/migrate/20240709072134_rename_terms_and_conditions_timestamp_to_terms_and_conditions_accepted_at.rb:
```diff
@@ -0,0 +1,6 @@
+class RenameTermsAndConditionsTimestampToTermsAndConditionsAcceptedAt < ActiveRecord::Migration[7.1]
+  def change
+    rename_column :dsi_users, :terms_and_conditions_timestamp, :terms_and_conditions_accepted_at
+
```
Unnecessary blank line.

```suggestion
```

felixclack reviewed (changes requested):
Left some comments to fix a few things up.

felixclack commented on config/feature_flags.yml:
```diff
@@ -36,3 +36,6 @@ feature_flags:
       them for a TRN when last name and date of birth return multiple results.
+  terms_and_conditions:
+    author: Richard Pattinson
+    description: Check whether a user has accepted the latest version of the terms and conditions.
```
Not needed anymore as you have removed the feature flag.

richardpattinson replied:
Updated

felixclack commented on spec/support/system/activate_features_steps.rb:
```diff
@@ -7,6 +7,10 @@ def given_the_qualifications_service_is_open
+  def given_the_terms_and_conditions_feature_flag_is_enabled
+    FeatureFlags::FeatureFlag.activate(:terms_and_conditions)
+  end
+
```
Not needed anymore

richardpattinson replied:
Updated
